### PR TITLE
feat(compile): traceable GDN decoder loop (Qwen3.5 + Qwen3-Next) + model-agnostic FA2 varlen collator

### PR DESCRIPTION
**Pull request by thad0ctor**
## Description

Under `torch_compile`, one graph break inside the GatedDeltaNet decoder for-loop made dynamo skip the whole text-model `forward` frame (a break inside a loop is unresumable), so every layer (attention included) ran eagerly.

This makes the loop traceable for **every GatedDeltaNet model** (Qwen3.5 dense + MoE, Qwen3-Next), plus a **model-agnostic** collator path that fixes the same break for *any* flash-attention + packed model under compile:

- **`gated_delta_net_ops.py` (new, shared)** — opaque `torch.library` ops (`axolotl_gdn::gdn_conv`/`gdn_chunk` + backward) wrapping the *same* FLA kernels with identical saved tensors. They derive `cu_seqlens` from `position_ids` **inside the op**, so no `aten.nonzero` enters the graph. A `cast_g` flag matches each architecture's eager call (qwen3_5 casts `g`; qwen3_next keeps f32) for **bitwise** parity; packed input enforces batch-1 with an explicit `ValueError`. Also hosts the shared `FusedRMSNormGated` compile boundary.
- **`qwen3_5/` + `qwen3_next/modeling.py`** — route the GDN training path through the ops under compile, and install the `FusedRMSNormGated` opaque wrapper (its backward recomputes via `autograd.grad`, since the FLA backward isn't meta-traceable). Self-attention stays in-graph under GC+training, behind a `dynamo.disable` boundary otherwise (guards an Inductor FA2-backward fusion that corrupts packed grads; pinned by `test_fa2_compiled_matches_eager_grads`). The qwen3_next wiring also fixes a pre-existing crash (its decoder passed `cache_position` to a `GatedDeltaNet.forward` that rejects it).
- **`collators/batching.py` + `builders/causal.py`** — under `torch_compile`, the multipack collator precomputes the FA2 varlen kwargs (`cu_seq_lens_q/k`, `max_length_q/k`) once per batch, so transformers skips its per-layer `nonzero` derivation. **Model-agnostic** — gated on `SUPPORTED_MULTIPACK_MODEL_TYPES`, so any packed FA2 model compiles clean.
- **`patch_manager.py`** — threads `torch_compile` into the qwen3_5/qwen3_5_moe/qwen3_next packing patches.

Opt-in via `torch_compile`; inert otherwise.

**Version coupling — the `==0.4.1` pin is load-bearing.** The ops wrap FLA *internal* functions by fixed arity and the fakes hardcode 0.4.1 shapes/dtypes; FLA 0.4.2/0.5.0 already changed `chunk_gated_delta_rule_fwd`'s arity (crash at first forward). So a bump needs revalidating the ops — guarded by `test_opcheck_custom_ops` + `test_eager_parity_ops_vs_legacy_bitwise` (the call-time crash isn't caught by the build-time warning). The upstream fix fla-org/flash-linear-attention#909 removes only the index-helper breaks, not the `nonzero`/`@torch.compiler.disable` ones these ops target, so the wrapper can't just be dropped on a newer FLA.

**Known limit:** auto-enabled LoRA triton kernels + a GDN packing patch + `torch_compile` + GC NaN gradients on **torch ≤ 2.10** (pre-existing on main; clean on 2.11, and with any one ingredient removed). `PatchManager` force-disables those kernels with a warning on torch < 2.11.

## Motivation and Context

The decoder loop is the bulk of training compute. Skipping its compilation forfeits the largest fusion payoff: the eager RMSNorm fp32 round-trips (~870 GB/step of traffic at 9B/32k, interposer-measured) + per-kernel Python launch overhead. Once the loop traces, Inductor fuses those.

Scope: the GDN ops cover **Qwen3.5/Qwen3.6** (`model_type` `qwen3_5`/`qwen3_5_moe`, dense + MoE, text + VL — every shipped checkpoint is `*ForConditionalGeneration`) and **Qwen3-Next** (`qwen3_next`, same FLA GatedDeltaNet kernels). The collator change is model-agnostic. Dense Qwen3/Qwen2.5/Qwen3-VL have no GatedDeltaNet, so the loop-internal breaks don't exist for them — but they still benefit from the collator path under FA2+packing+compile.

## How has this been tested?

**Stack:** torch 2.11.0+cu130, transformers 5.9.0, FLA 0.4.1, liger-kernel 0.8.0, real flash_attn 2.8.3 (the validated stack). RTX 3090 / 3090 Ti / 5090.

**Functional — 12 tests, all passing:**

`tests/monkeypatch/test_qwen3_5_compiled_loop.py` (10): zero-breaks-with-GC (1 graph), FA2-loop-with-varlen-kwargs, VL `ForConditionalGeneration` 3-D MRoPE, `aten.nonzero`-break-gone, **bitwise eager parity** (ops vs legacy), compiled-vs-eager, **FA2+GC compiled grad parity**, toy MoE (0 breaks), B>1-packed-raises, `opcheck` (packed/MRoPE/dense × bias × both `cast_g`).
`tests/monkeypatch/test_qwen3_next_compiled_loop.py` (2): qwen3_next decoder loop **0 breaks** + **bitwise eager parity** (`cast_g=False`, g stays f32).
Existing `test_qwen3_5_fused_attn.py`: 12/12 (no regression). Suite also passes one major version down (torch 2.9.1 / 2.10.0 + transformers 5.5.4).

Multimodal: Multimodal training is non-packed in axolotl (SFT vision examples + MM-CPT (pending PR #3629) ship `sample_packing: false`), so the GDN ops are inert there by construction this was validated with real images (llava-instruct through the Qwen3.5-2B vision tower: packing patches absent from logs, eager-vs-compiled step-1 loss Δ0.004, eval clean).

Multiple test training runs and benches (below).

---

### Benchmarks (all on the validated stack above)

Qwen3.5-2B/9B = `Qwen3_5ForConditionalGeneration`, LoRA r=16 q/k/v/o, sample packing, bf16, gradient checkpointing, `attn_implementation: flash_attention_2`. 60 steps, steady-state window after warmup, `TORCH_LOGS=recompiles`. Throughput = tok/s; VRAM = reserved (GiB). **Zero NaN and step-1/final-10 loss parity in every cell.**

**A. Core compile speedup** On real FA2, `torch_compile` *without* this PR is near-neutral (the loop graph-breaks); this PR is what makes compile worth it (RTX 3090):

| model | eager | compile (main, no PR) | compile **+ PR** | PR vs eager | PR vs compile-main |
|---|---|---|---|---|---|
| Qwen3.5-2B (seq 4096) | 3593 tok/s | 3788 (+5.4%) | **4617** | **+28.5%** | **+21.9%** |
| Qwen3.5-9B (seq 4096, CCE) | 2724 tok/s | 2732 (+0.3%) | **3330** | **+22.2%** | **+21.9%** |

**B. Feature compatibility / composition (Qwen3.5-2B, compile + PR, varying feature):** (RTX 3090)

| config | tok/s | reserved VRAM | note |
|---|---|---|---|
| compile + PR (LoRA kernels on — default) | 4617 | 14.66 GiB | baseline |
| + LoRA kernels **off** | 4585 | 14.40 GiB | kernels ≈ neutral under compile |
| + `fused_attn_kernel` (#3680) | 4394 | 14.67 GiB | −4.8% — redundant under compile |
| + axolotl **CCE** (`cut_cross_entropy`) | **4781** | **6.03 GiB** | fastest **and** ~2.4× less memory |
| + liger `fused_linear_cross_entropy` | 4593 | 14.66 GiB | **no-op** on Qwen3.5 (= no memory benefit) — requires overlooked gating bug fix following liger 0.8.0 bump |
| + liger full stack (rms_norm/gated/swiglu/rope) | 4137 | 10.60 GiB | −10% — kernels redundant under compile |
| *(eager + `fused_attn_kernel`)* | 3648 | 18.07 GiB | eager: fused_attn only +1.5% |

LoRA triton kernels and `fused_attn_kernel` are subsumed by Inductor once the loop compiles (consistent with #3680's own "fused+compile −18%" dense-Qwen3 numbers); liger composes cleanly but adds no value under compile. 

**C. Generalization — the model-agnostic collator (compile on/off, FA2 + packing):** (RTX 5090)

| model | eager | compile **+ PR** | speedup | reserved (eager → compile) |
|---|---|---|---|---|
| **Llama-3.2-1B** (seq 2048) | 5181 tok/s | **7572** | **+46.1%** | 6.09 → 4.54 GiB |
| **Qwen3-0.6B** (seq 2048) | 5445 tok/s | **11375** | **+108.9%** | 5.54 → 5.01 GiB |

Neither model has a GatedDeltaNet so  speedups come  from the collator removing the FA2 per-layer break. Isolation (toy Llama, FA2 + packed, under compile): **3 graph breaks (4 graphs) without the varlen kwargs → 0 breaks (1 graph) with them.** 0 recompiles in steady state; loss parity holds.

---

**Convergence:** loss is bit-stable across every cell above — step-1 identical to eager and final-10 within float noise (e.g. 2B eager 1.1275 vs compile+PR 1.1275; 9B 0.9264 vs 0.9266). The opaque-op path is *bitwise* identical to legacy eager (regression-tested) for both qwen3_5 and qwen3_next.

**Composition with axolotl-ai-cloud/axolotl#3732 (fused-LoRA GDN routing):** merges cleanly, both touch the GDN forward, the projection call sites route through `_la_proj_fwd` while the compiled-ops gating stays intact, and the fused `LoRA_O` autograd Function traces inside the compiled loop with 0 graph breaks (earlier 2B bench: both-merged+compile +44% vs eager, #3732's +10% marginal surviving compilation because it substitutes bf16 adapter GEMMs Inductor can't infer).

## AI Usage Disclaimer
Opus 4.8 / Fable 5 used throughout.

## Types of changes
- [x] New feature (`torch_compile` support for the GatedDeltaNet decoder loop — Qwen3.5 dense/MoE + Qwen3-Next)
- [x] New feature (model-agnostic FA2 varlen collator kwargs for any packed model under compile)
- [x] Performance improvement (non-breaking; numerics-neutral, bitwise eager fallback)
- [x] New tests

## Release Notes

* **New Features**
  * Added `torch.compile` support for Qwen3.5/Qwen3-Next models with optimized compiled execution paths
  * Enhanced FlashAttention v2 varlen kwargs precomputation for improved performance
  * Implemented custom compiled operators for kernel optimization

* **Bug Fixes**
  * Fixed NaN gradients in LoRA kernels for specific Qwen3.5 configurations
  * Eliminated unwanted graph breaks in compiled execution paths

* **Tests**
  * Added comprehensive CUDA regression tests for compiled decoder loops with eager/compiled parity validation

### CODE REVIEW

File: tests/monkeypatch/test_qwen3_5_compiled_loop.py [hunk 1]
```diff
@@ -0,0 +1,458 @@
...
+        torch.manual_seed(123)
+        T = 64
+        input_ids = torch.randint(0, 128, (2, T), device="cuda")
+        pos = torch.cat(
+            [torch.arange(40, device="cuda"), torch.arange(T - 40, device="cuda")]
+        )
+        position_ids = pos.view(1, T).expand(2, T).contiguous()
+
+        with pytest.raises(Exception, match="batch size is expected to be 1"):
+            model(input_ids=input_ids, position_ids=position_ids, use_cache=False)
```
[the remark this answers is not in the document]
Comment (thad0ctor) [reply tests/monkeypatch/test_qwen3_5_compiled_loop.py]: Addressed with [31b8d94](https://github.com/axolotl-ai-cloud/axolotl/pull/3734/commits/31b8d9478821d136cd5c87543e4f3358ab11656e)

File: tests/monkeypatch/test_qwen3_5_compiled_loop.py [hunk 2]
```diff
@@ -0,0 +1,458 @@
...
+    yield qm
+    hf.Qwen3_5DecoderLayer.forward = saved["decoder_forward"]
+    hf.Qwen3_5GatedDeltaNet.forward = saved["gdn_forward"]
+    FusedRMSNormGated.forward = saved["norm_forward"]
+    if saved["norm_flag"]:
+        FusedRMSNormGated._axolotl_compile_boundary = saved["norm_flag"]
+    else:
+        try:
+            delattr(FusedRMSNormGated, "_axolotl_compile_boundary")
+        except AttributeError:
```
[the remark this answers is not in the document]
Comment (thad0ctor) [reply tests/monkeypatch/test_qwen3_5_compiled_loop.py]: Addressed with [31b8d94](https://github.com/axolotl-ai-cloud/axolotl/pull/3734/commits/31b8d9478821d136cd5c87543e4f3358ab11656e)

File: src/axolotl/utils/collators/batching.py [hunk 3]
```diff
@@ -122,6 +127,28 @@ def __call__(self, features, return_tensors=None):
...
+                features["cu_seq_lens_k"] = cu_k
+                features["max_length_q"] = int(max_q)
+                features["max_length_k"] = int(max_k)
+            except (
+                Exception
+            ) as exc:  # pragma: no cover - fall back to per-layer derivation
+                LOG.warning_once(
+                    f"FA varlen kwargs precompute failed ({exc}); falling back to "
+                    "transformers' per-layer derivation (decoder loop will not compile)"
+                )
```
[the remark this answers is not in the document]
Comment (thad0ctor) [reply src/axolotl/utils/collators/batching.py:155]: Addressed with [31b8d94](https://github.com/axolotl-ai-cloud/axolotl/pull/3734/commits/31b8d9478821d136cd5c87543e4f3358ab11656e)

File: src/axolotl/monkeypatch/models/qwen3_5/modeling.py [hunk 4]
```diff
@@ -20,6 +20,29 @@
...
+    return attn_module(**kwargs)
+
+
+try:
+    import torch._dynamo as _dynamo
+
+    _call_self_attn_disabled = _dynamo.disable(_call_self_attn)
+except Exception:  # pragma: no cover
+    _call_self_attn_disabled = _call_self_attn
+
```
Comment (ved1beta) [inline src/axolotl/monkeypatch/models/qwen3_5/modeling.py]: this block is twice for both files , make it one and import

**Review by ved1beta — CHANGES_REQUESTED** [review]

File: src/axolotl/monkeypatch/models/qwen3_5/modeling.py (hunk 4 above)
Comment (thad0ctor) [reply src/axolotl/monkeypatch/models/qwen3_5/modeling.py]: Addressed via [c7dd43e](https://github.com/axolotl-ai-cloud/axolotl/pull/3734/commits/c7dd43ea5ef84cac45d938e0d0bf4dbce470d79f)

**Comment by thad0ctor** [comment]
@ved1beta addressed your comments via [c7dd43e](https://github.com/axolotl-ai-cloud/axolotl/pull/3734/commits/c7dd43ea5ef84cac45d938e0d0bf4dbce470d79f)

File: src/axolotl/monkeypatch/models/fla_compiled_loop.py [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+def _call_self_attn(attn_module, **kwargs):
+    return attn_module(**kwargs)
+
+
+try:
+    import torch._dynamo as _dynamo
+
+    call_self_attn_disabled = _dynamo.disable(_call_self_attn)
+except Exception:  # pragma: no cover
+    call_self_attn_disabled = _call_self_attn
```
Comment (ved1beta) [inline src/axolotl/monkeypatch/models/fla_compiled_loop.py:19]: dont need a new file here

**Comment by thad0ctor** [comment]
Resolved merge conflicts [82d3b8b](https://github.com/axolotl-ai-cloud/axolotl/pull/3734/commits/82d3b8bb06441552478f7290f2d79f50ebe32fcb)

**Review by winglian — APPROVED** [review]

**Comment by winglian** [comment]
@thad0ctor any preference on merge order? #3869 first and then refactor this? or merge this first and then clean/refactor these model changes into the model support? My preference is to merge the ModelSupport one first for a cleaner history

**Comment by thad0ctor** [comment]
> @thad0ctor any preference on merge order? #3869 first and then refactor this? or merge this first and then clean/refactor these model changes into the model support? My preference is to merge the ModelSupport one first for a cleaner history

Up to you! Make perfect sense

**Comment by thad0ctor** [comment]
> @thad0ctor any preference on merge order? #3869 first and then refactor this? or merge this first and then clean/refactor these model changes into the model support? My preference is to merge the ModelSupport one first for a cleaner history

@winglian as part of the refactor, would you prefer a minimal rebase off of https://github.com/axolotl-ai-cloud/axolotl/pull/3869 as well as a allignment of qwen3 into the model support (whether as a precursor PR or rolled into this)?

**Review by ved1beta — APPROVED** [review]
